enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,342 @@
+  - "@derekwaynecarr"
+creation-date: 2020-06-25
+last-updated: 2020-06-25
+status: provisional
+see-also:
+  - /enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md
+  - https://github.com/openshift/enhancements/pull/384
+replaces: []
+superseded-by: []
+---
+
+# Subscription Injection Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+      [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will allow any workload to access RHEL subscription content via a pod annotation.
+
+## Motivation
+
+Containers can download RHEL content if their hosts have subscriptions attached. In 3.x, all nodes
+were assumed to be RHEL 7 nodes that were subscribed individually. The information needed to access
+subscription content (entitlement keys and `redhat.repo` configurations) was symbolically linked
+into the running container at a well-known location that yum/dnf could find. This capability
+existed in our patch of Docker, and was carried over into Red Hat’s container toolchain used on
+OpenShift 4 (cri-o, podman, buildah, etc.).
+
+In OpenShift 4 the preferred operating system (RHCOS) is not capable of attaching subscriptions
+individually. To access subscription content, containers must be provided entitlement keys and
+optionally `redhat.repo` configurations via other means. Enabling this capability needs to be
+simple and straightforward.
+
+### Goals
+
+1. Inject entitlement keys, such as the Simple Content Access key, into a pod spec.
+2. Allow overrides of the `redhat.repo` configuration for containers that need to connect to
+   Satellite.
+3. Define APIs and conventions that allow cluster admins to add the required information to a
+   cluster (entitlement keys and `redhat.repo` configurations).
+
+### Non-Goals
+
+1. Automatically install entitlement keys and `redhat.repo` configurations onto a subscribed
+   cluster.
+
+## Proposal
+
+### User Stories
+
+#### Add RHEL content to running containers
+
+As a developer or application administrator
+I want access to RHEL subscription content in my containers
+So that I can install RHEL content within my container via `yum install`
+
+#### Access RHEL content from a Satellite instance
+
+As a developer using OpenShift 
+I want to be able to access RHEL content from my Satellite instance
+So that I download RHEL content from my Satellite instances instead of the Red Hat CDN.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Installation and Setup
+
+When the `subscription-injection-operator` is installed, it creates a subscription `ClusterBundle`
+custom resource definition. The operator deployment creates a default instance which is
+cluster-scoped:
+
+```yaml
+kind: ClusterBundle
```

Comment:
why is this a clusterbundle and not a subscriptionbundle?  trying to avoid semi-stuttering?

Reply:
(could we just call it a bundle?  bundle.subscription.openshift.io probably provides sufficient context?

Reply:
Multi-tenancy was a concern when I renamed the API to `ClusterBundle`. Adding a `Bundle` API that is namespace scoped is something I can add to the proposal, and is easier to implement (won't need the projected resource csi driver, we can use plain old `Secret`/`ConfigMap` volume mounts).

Reply:
hmmmm.  i mean we can't drop the "i put the creds in one space and everyone in my cluster gets them" functionality.

in my head multitenant/multinamespace means:

"I can put the bundle in a namespace i control, and expose it to the users i choose.  Those users can use it in any namespace they can create pods in"

Reply:
Single namespace is straightforward - I create a bundle in my namespace, with secrets/ConfigMaps in the same namespace.

Sharing across multiple namespaces implies we use a `ClusterBundle`, but then admins use RBAC to limit who gets what. Ex: `ClusterBundle` creates a `ClusterRole`, admins then add cluster roles to users for the desired namespaces.